File: def_hierarchical_pathfinding/src/local_pathfinder.lua
-- Local pathfinding within a single chunk using A* algorithm
-- Finds paths on water tiles (value 0) avoiding land tiles (value 1)

local data_structures = require "src.utils.data_structures"

local M = {}

-- Check if position is walkable (water tile)
-- @param chunk_data 2D array of tiles
-- @param pos Position {x, y}
-- @return boolean
function M.is_walkable(chunk_data, pos)
    if pos.x < 0 or pos.y < 0 then
        return false
    end
    
    local row = chunk_data[pos.y + 1]  -- Lua arrays are 1-indexed
    if not row then
        return false
    end
    
    local tile = row[pos.x + 1]
    return tile == 0  -- 0 = water (walkable)
end

-- Get valid neighbors for 4-directional movement
-- @param chunk_data 2D array of tiles
-- @param pos Current position {x, y}
-- @return Array of neighbor positions
function M.get_neighbors(chunk_data, pos)
    local neighbors = {}
    local directions = {
        {x = 0, y = -1},  -- Up
        {x = 0, y = 1},   -- Down
        {x = -1, y = 0},  -- Left
        {x = 1, y = 0}    -- Right
    }
    
    for _, dir in ipairs(directions) do
        local neighbor = {x = pos.x + dir.x, y = pos.y + dir.y}
        if M.is_walkable(chunk_data, neighbor) then
            table.insert(neighbors, neighbor)
        end
    end
    
    return neighbors
end

-- Calculate heuristic distance (Manhattan by default)
-- @param a Position {x, y}
-- @param b Position {x, y}
-- @param heuristic "manhattan" or "euclidean"
-- @return Distance
local function calculate_heuristic(a, b, heuristic)
    if heuristic == "euclidean" then
        local dx = b.x - a.x
        local dy = b.y - a.y
        return math.sqrt(dx * dx + dy * dy)
    else  -- manhattan (default)
        return math.abs(b.x - a.x) + math.abs(b.y - a.y)
    end
end

-- Find path using A* algorithm
-- @param chunk_data 2D array of tiles (0 = water, 1 = land)
-- @param start_pos Start position {x, y} in local coordinates
-- @param end_pos End position {x, y} in local coordinates
-- @param config Optional configuration {heuristic = "manhattan"|"euclidean", optimize_path = boolean}
-- @return Array of positions forming path, or nil if no path exists
function M.find_path(chunk_data, start_pos, end_pos, config)
    config = config or {}
    local heuristic = config.heuristic or "manhattan"
    
    -- Validate positions
    if start_pos.x < 0 or start_pos.y < 0 or end_pos.x < 0 or end_pos.y < 0 then
        error("Invalid position: negative coordinates")
    end
    
    local chunk_size = #chunk_data
    if chunk_size == 0 or #chunk_data[1] == 0 then
        error("Invalid chunk data: empty")
    end
    
    if start_pos.x >= chunk_size or start_pos.y >= chunk_size or
       end_pos.x >= chunk_size or end_pos.y >= chunk_size then
        error("Position outside chunk bounds")
    end
    
    -- Check if start and end are walkable
    if not M.is_walkable(chunk_data, start_pos) or not M.is_walkable(chunk_data, end_pos) then
        return nil
    end
    
    -- Handle same position
    if start_pos.x == end_pos.x and start_pos.y == end_pos.y then
        return {start_pos}
    end
    
    -- A* implementation
    local PriorityQueue = data_structures.PriorityQueue
    local open_set = PriorityQueue:new()
    local came_from = {}
    local g_score = {}
    local closed_set = {}
    
    -- Helper to create position key
    local function pos_key(pos)
        return pos.x .. "," .. pos.y
    end
    
    -- Initialize start node
    local start_key = pos_key(start_pos)
    g_score[start_key] = 0
    
    open_set:push({
        pos = start_pos,
        g_score = 0,
        f_score = calculate_heuristic(start_pos, end_pos, heuristic),
        priority = calculate_heuristic(start_pos, end_pos, heuristic)
    })
    
    -- Main A* loop
    while not open_set:empty() do
        local current = open_set:pop()
        local current_key = pos_key(current.pos)
        
        -- Check if we reached the goal
        if current.pos.x == end_pos.x and current.pos.y == end_pos.y then
            -- Reconstruct path
            local path = {}
            local pos = end_pos
            
            while pos do
                table.insert(path, 1, {x = pos.x, y = pos.y})
                local key = pos_key(pos)
                pos = came_from[key]
            end
            
            -- Optimize path if requested
            if config.optimize_path then
                path = M.optimize_path(path, chunk_data)
            end
            
            return path
        end
        
        closed_set[current_key] = true
        
        -- Examine neighbors
        local neighbors = M.get_neighbors(chunk_data, current.pos)
        
        for _, neighbor in ipairs(neighbors) do
            local neighbor_key = pos_key(neighbor)
            
            if not closed_set[neighbor_key] then
                local tentative_g = current.g_score + 1  -- Cost is 1 for adjacent tiles
                
                if not g_score[neighbor_key] or tentative_g < g_score[neighbor_key] then
                    -- Update scores
                    came_from[neighbor_key] = current.pos
                    g_score[neighbor_key] = tentative_g
                    local f_score = tentative_g + calculate_heuristic(neighbor, end_pos, heuristic)
                    
                    -- Add to open set
                    open_set:push({
                        pos = neighbor,
                        g_score = tentative_g,
                        f_score = f_score,
                        priority = f_score
                    })
                end
            end
        end
    end
    
    -- No path found
    return nil
end

-- Optimize path by removing unnecessary waypoints
-- @param path Array of positions
-- @param chunk_data Chunk data for line-of-sight checks
-- @return Optimized path
function M.optimize_path(path, chunk_data)
    if #path <= 2 then
        return path
    end
    
    local optimized = {path[1]}
    local current_index = 1
    
    while current_index < #path do
        local furthest_visible = current_index + 1
        
        -- Find furthest point we can see in straight line
        for i = current_index + 2, #path do
            if M.has_line_of_sight(path[current_index], path[i], chunk_data) then
                furthest_visible = i
            else
                break
            end
        end
        
        table.insert(optimized, path[furthest_visible])
        current_index = furthest_visible
    end
    
    return optimized
end

-- Check if there's a clear line of sight between two points
-- @param from Start position
-- @param to End position
-- @param chunk_data Chunk data
-- @return boolean
function M.has_line_of_sight(from, to, chunk_data)
    -- Bresenham's line algorithm
    local x0, y0 = from.x, from.y
    local x1, y1 = to.x, to.y
    
    local dx = math.abs(x1 - x0)
    local dy = math.abs(y1 - y0)
    local sx = x0 < x1 and 1 or -1
    local sy = y0 < y1 and 1 or -1
    local err = dx - dy
    
    while true do
        if not M.is_walkable(chunk_data, {x = x0, y = y0}) then
            return false
        end
        
        if x0 == x1 and y0 == y1 then
            break
        end
        
        local e2 = 2 * err
        if e2 > -dy then
            err = err - dy
            x0 = x0 + sx
        end
        if e2 < dx then
            err = err + dx
            y0 = y0 + sy
        end
    end
    
    return true
end

return M 

File: library_example/hierarchical_pathfinding/src/LocalPathfinder.js
/**
 * Local pathfinding within a single chunk using A* and JPS
 * Finds paths on walkable tiles (0) avoiding blocked tiles (1)
 */

import { CoordUtils } from './utils/CoordUtils.js';

export class LocalPathfinder {
    /**
     * Check if position is walkable
     * @param {Array} chunkData - 2D array of tiles
     * @param {Object} pos - Position {x, y}
     * @returns {boolean}
     */
    static isWalkable(chunkData, pos) {
        // Check boundaries
        if (pos.x < 0 || pos.y < 0 || 
            pos.y >= chunkData.length || 
            pos.x >= chunkData[0].length) {
            return false;
        }
        
        // 0 = walkable, 1 = blocked
        return chunkData[pos.y][pos.x] === 0;
    }

    /**
     * Find path using specified algorithm
     * @param {Array} chunkData - 2D array of tiles (0=walkable, 1=blocked)
     * @param {Object} startPos - Start {x, y}
     * @param {Object} endPos - End {x, y}
     * @param {string} algorithm - Algorithm to use ('astar' or 'jps')
     * @param {string} heuristic - Heuristic to use
     * @param {number} heuristicWeight - Weight for heuristic (default 1.0)
     * @returns {Array|null} - Array of positions or null
     */
    static findPath(chunkData, startPos, endPos, algorithm = 'astar', heuristic = 'manhattan', heuristicWeight = 1.0) {
        // Check if start and end are walkable
        if (!this.isWalkable(chunkData, startPos) || 
            !this.isWalkable(chunkData, endPos)) {
            return null;
        }

        // Same position
        if (startPos.x === endPos.x && startPos.y === endPos.y) {
            return [startPos];
        }

        // Choose algorithm
        switch (algorithm) {
            case 'jps':
                return this.findPathJPS(chunkData, startPos, endPos, heuristic, heuristicWeight);
            case 'astar':
            default:
                return this.findPathAStar(chunkData, startPos, endPos, heuristic, heuristicWeight);
        }
    }

    /**
     * Find path using A* algorithm
     * @param {Array} chunkData - 2D array of tiles (0=walkable, 1=blocked)
     * @param {Object} startPos - Start {x, y}
     * @param {Object} endPos - End {x, y}
     * @param {string} heuristic - Heuristic to use
     * @param {number} heuristicWeight - Weight for heuristic
     * @returns {Array|null} - Array of positions or null
     */
    static findPathAStar(chunkData, startPos, endPos, heuristic = 'manhattan', heuristicWeight = 1.0) {
        // Structures for A*
        const openList = [];
        const closedSet = new Set();
        const cameFrom = {};
        const gScore = {};
        const fScore = {};

        // Helper for keys
        const key = (pos) => `${pos.x},${pos.y}`;
        
        // Get heuristic function
        const heuristicFunc = CoordUtils.getHeuristic(heuristic);

        // Initialize start
        const startKey = key(startPos);
        gScore[startKey] = 0;
        fScore[startKey] = heuristicFunc(startPos, endPos) * heuristicWeight;
        openList.push({ pos: startPos, f: fScore[startKey] });

        // Main A* loop
        while (openList.length > 0) {
            // Find node with lowest f-score
            openList.sort((a, b) => a.f - b.f);
            const current = openList.shift();
            const currentKey = key(current.pos);

            // Found the goal!
            if (current.pos.x === endPos.x && current.pos.y === endPos.y) {
                return this.reconstructPath(cameFrom, endPos);
            }

            closedSet.add(currentKey);

            // Check neighbors (4 directions)
            const neighbors = [
                { x: current.pos.x, y: current.pos.y - 1 }, // Up
                { x: current.pos.x, y: current.pos.y + 1 }, // Down
                { x: current.pos.x - 1, y: current.pos.y }, // Left
                { x: current.pos.x + 1, y: current.pos.y }   // Right
            ];

            for (const neighbor of neighbors) {
                // Skip unwalkable
                if (!this.isWalkable(chunkData, neighbor)) continue;

                const neighborKey = key(neighbor);
                
                // Skip already visited
                if (closedSet.has(neighborKey)) continue;

                // Calculate cost
                const tentativeG = gScore[currentKey] + 1;

                // Check if this is a better path
                if (!(neighborKey in gScore) || tentativeG < gScore[neighborKey]) {
                    // Save path
                    cameFrom[neighborKey] = current.pos;
                    gScore[neighborKey] = tentativeG;
                    fScore[neighborKey] = tentativeG + heuristicFunc(neighbor, endPos) * heuristicWeight;

                    // Add to open list if not already there
                    if (!openList.some(n => n.pos.x === neighbor.x && n.pos.y === neighbor.y)) {
                        openList.push({ pos: neighbor, f: fScore[neighborKey] });
                    }
                }
            }
        }

        // No path found
        return null;
    }

    /**
     * Find path using JPS (Jump Point Search) algorithm
     * @param {Array} chunkData - 2D array of tiles (0=walkable, 1=blocked)
     * @param {Object} startPos - Start {x, y}
     * @param {Object} endPos - End {x, y}
     * @param {string} heuristic - Heuristic to use
     * @param {number} heuristicWeight - Weight for heuristic
     * @returns {Array|null} - Array of positions or null
     */
    static findPathJPS(chunkData, startPos, endPos, heuristic = 'manhattan', heuristicWeight = 1.0) {
        // Structures for JPS
        const openList = [];
        const closedSet = new Set();
        const cameFrom = {};
        const gScore = {};
        const fScore = {};

        // Helper for keys
        const key = (pos) => `${pos.x},${pos.y}`;
        
        // Get heuristic function
        const heuristicFunc = CoordUtils.getHeuristic(heuristic);

        // Initialize start
        const startKey = key(startPos);
        gScore[startKey] = 0;
        fScore[startKey] = heuristicFunc(startPos, endPos) * heuristicWeight;
        openList.push({ pos: startPos, f: fScore[startKey] });

        // Main JPS loop
        while (openList.length > 0) {
            // Find node with lowest f-score
            openList.sort((a, b) => a.f - b.f);
            const current = openList.shift();
            const currentKey = key(current.pos);

            // Found the goal!
            if (current.pos.x === endPos.x && current.pos.y === endPos.y) {
                return this.reconstructPath(cameFrom, endPos);
            }

            closedSet.add(currentKey);

            // Get jump points for current node
            const jumpPoints = this.getJumpPoints(chunkData, current.pos, endPos);

            for (const jumpPoint of jumpPoints) {
                const jumpKey = key(jumpPoint.pos);
                
                // Skip already visited
                if (closedSet.has(jumpKey)) continue;

                // Calculate cost (distance to jump point)
                const distance = this.calculateDistance(current.pos, jumpPoint.pos);
                const tentativeG = gScore[currentKey] + distance;

                // Check if this is a better path
                if (!(jumpKey in gScore) || tentativeG < gScore[jumpKey]) {
                    // Save path
                    cameFrom[jumpKey] = current.pos;
                    gScore[jumpKey] = tentativeG;
                    fScore[jumpKey] = tentativeG + heuristicFunc(jumpPoint.pos, endPos) * heuristicWeight;

                    // Add to open list if not already there
                    if (!openList.some(n => n.pos.x === jumpPoint.pos.x && n.pos.y === jumpPoint.pos.y)) {
                        openList.push({ pos: jumpPoint.pos, f: fScore[jumpKey] });
                    }
                }
            }
        }

        // No path found
        return null;
    }

    /**
     * Get jump points for JPS algorithm
     * @param {Array} chunkData - 2D array of tiles
     * @param {Object} current - Current position
     * @param {Object} goal - Goal position
     * @returns {Array} - Array of jump points with positions and distances
     */
    static getJumpPoints(chunkData, current, goal) {
        const jumpPoints = [];
        
        // 8 directions for JPS
        const directions = [
            { dx: 0, dy: -1 },  // North
            { dx: 1, dy: -1 },  // Northeast
            { dx: 1, dy: 0 },   // East
            { dx: 1, dy: 1 },   // Southeast
            { dx: 0, dy: 1 },   // South
            { dx: -1, dy: 1 },  // Southwest
            { dx: -1, dy: 0 },  // West
            { dx: -1, dy: -1 }  // Northwest
        ];

        for (const dir of directions) {
            const jumpPoint = this.findJumpPoint(chunkData, current, dir, goal);
            if (jumpPoint) {
                jumpPoints.push(jumpPoint);
            }
        }

        return jumpPoints;
    }

    /**
     * Find jump point in given direction
     * @param {Array} chunkData - 2D array of tiles
     * @param {Object} start - Start position
     * @param {Object} direction - Direction {dx, dy}
     * @param {Object} goal - Goal position
     * @returns {Object|null} - Jump point with position and distance
     */
    static findJumpPoint(chunkData, start, direction, goal) {
        const { dx, dy } = direction;
        let current = { x: start.x + dx, y: start.y + dy };
        let distance = 1;

        // Check if initial step is walkable
        if (!this.isWalkable(chunkData, current)) {
            return null;
        }

        // Check if goal is reached
        if (current.x === goal.x && current.y === goal.y) {
            return { pos: current, distance };
        }

        // For diagonal movement, check if both cardinal directions are walkable
        if (dx !== 0 && dy !== 0) {
            const horizontalWalkable = this.isWalkable(chunkData, { x: start.x + dx, y: start.y });
            const verticalWalkable = this.isWalkable(chunkData, { x: start.x, y: start.y + dy });
            
            if (!horizontalWalkable || !verticalWalkable) {
                return null;
            }
        }

        // Jump in the direction until we hit an obstacle or find a jump point
        while (this.isWalkable(chunkData, current)) {
            // Check if this is a jump point (has forced neighbors)
            if (this.hasForcedNeighbors(chunkData, current, direction)) {
                return { pos: current, distance };
            }

            // Check if goal is reached
            if (current.x === goal.x && current.y === goal.y) {
                return { pos: current, distance };
            }

            // For diagonal movement, try to jump in cardinal directions
            if (dx !== 0 && dy !== 0) {
                // Try horizontal jump
                const horizontalJump = this.findJumpPoint(chunkData, current, { dx, dy: 0 }, goal);
                if (horizontalJump) {
                    return { pos: current, distance };
                }

                // Try vertical jump
                const verticalJump = this.findJumpPoint(chunkData, current, { dx: 0, dy }, goal);
                if (verticalJump) {
                    return { pos: current, distance };
                }
            }

            // Move to next position
            current = { x: current.x + dx, y: current.y + dy };
            distance++;
        }

        return null;
    }

    /**
     * Check if position has forced neighbors (simplified version)
     * @param {Array} chunkData - 2D array of tiles
     * @param {Object} pos - Position to check
     * @param {Object} direction - Current direction
     * @returns {boolean} - True if has forced neighbors
     */
    static hasForcedNeighbors(chunkData, pos, direction) {
        // Simplified forced neighbor check
        // In a full JPS implementation, this would check for specific patterns
        // For now, we'll consider any position with diagonal movement as a potential jump point
        return direction.dx !== 0 && direction.dy !== 0;
    }

    /**
     * Calculate distance between two positions
     * @param {Object} a - First position
     * @param {Object} b - Second position
     * @returns {number} - Distance
     */
    static calculateDistance(a, b) {
        const dx = Math.abs(b.x - a.x);
        const dy = Math.abs(b.y - a.y);
        
        if (dx === dy) {
            return dx * Math.SQRT2; // Diagonal movement
        } else {
            return dx + dy; // Manhattan distance
        }
    }

    /**
     * Reconstruct path from predecessors map
     * @param {Object} cameFrom - Predecessors map
     * @param {Object} endPos - End position
     * @returns {Array} - Path of positions
     */
    static reconstructPath(cameFrom, endPos) {
        const path = [];
        let current = endPos;
        
        while (current) {
            path.unshift({ x: current.x, y: current.y });
            const key = `${current.x},${current.y}`;
            current = cameFrom[key];
        }
        
        return path;
    }
} 